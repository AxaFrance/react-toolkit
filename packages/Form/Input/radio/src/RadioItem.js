import React from 'react';
import PropTypes from 'prop-types';

import { InputList, InputConstants as Constants, withInput, omit } from '@axa-fr/react-toolkit-form-core';
import { InputManager } from '@axa-fr/react-toolkit-core';

const omitProperties = omit([
  'classModifier',
  'className',
  'isVisible',
  'helpMessage',
  'options'
]);

const RadioItem = ({
                     optionClassName,
                     disabled,
                     value,
                     isChecked,
                     id,
                     name,
                     inputRef,
                     readOnly,
                     onChange,
                     onBlur,
                     onFocus,
                     children,
                     label,
                     ...otherProps
                   }) => {
  const newLabel = children || label;
  const newId = InputManager.getInputId(id); // id is require on this component
  return (
    <div className={optionClassName} key={value}>
      <input
        className="af-form__input-radio"
        name={name}
        id={newId}
        type="radio"
        value={value}
        onChange={onChange}
        onBlur={onBlur}
        onFocus={onFocus}
        checked={isChecked}
        readOnly={readOnly}
        disabled={disabled}
        ref={inputRef}
        {...omitProperties(otherProps)}
      />
      <label className="af-form__label" htmlFor={newId}>
        <span className="af-form__description">{newLabel}</span>
      </label>
    </div>
  );
};

const propTypes = {
  ...Constants.propTypes,
  value: PropTypes.string,
  label: PropTypes.oneOfType([PropTypes.element, PropTypes.string]),
  isChecked: PropTypes.bool,
};
const defaultClassName = 'af-form__radio';
const defaultProps = {
  ...Constants.defaultProps,
  value: '',
  className: defaultClassName,
  classModifier: '',
  isChecked: false,
};

const props = ({ className, classModifier, disabled }) => ({
  optionClassName: InputList.getOptionClassName(
    className,
    classModifier,
    defaultClassName,
    disabled
  ),
});

const EnhancedComponent = withInput(
  defaultClassName,
  propTypes,
  defaultProps,
  {},
  props
)(RadioItem);
EnhancedComponent.displayName = RadioItem.name;

export default EnhancedComponent;
