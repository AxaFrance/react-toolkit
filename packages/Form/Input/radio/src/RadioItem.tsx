import React, { InputHTMLAttributes } from 'react';
import { InputManager, ClassManager } from '@axa-fr/react-toolkit-core';
import {
  CustomFormEvent,
  defaultOnChange,
} from '@axa-fr/react-toolkit-form-core';

export interface RadioItemProps
  extends Omit<InputHTMLAttributes<HTMLInputElement>, 'onChange'> {
  optionClassName?: string;
  inputRef?: React.LegacyRef<HTMLInputElement>;
  label: string;
  onChange?: (event: CustomFormEvent) => void;
  classModifier?: string;
}

const RadioItem: React.FC<RadioItemProps> = ({
  className = defaultClassName,
  disabled,
  value = '',
  id,
  inputRef,
  readOnly,
  onChange,
  children,
  label,
  classModifier = '',
  optionClassName,
  ...otherProps
}) => {
  const newLabel = children || label;
  const newId = InputManager.getInputId(id); // id is required on this component
  const classModifierWithDisabled = `${classModifier}${
    disabled ? ' disabled' : ''
  }`;
  const divOptionClassName = ClassManager.getComponentClassName(
    className,
    classModifierWithDisabled,
    defaultClassName
  );
  return (
    <div className={`${divOptionClassName} ${classModifier}`}>
      <input
        className="af-form__input-radio"
        value={value}
        id={newId}
        type="radio"
        onChange={defaultOnChange(onChange)}
        ref={inputRef}
        disabled={disabled}
        {...otherProps}
      />
      <label className="af-form__label" htmlFor={newId}>
        <span className="af-form__description">{newLabel}</span>
      </label>
    </div>
  );
};

export const defaultClassName = 'af-form__radio';

export default RadioItem;
