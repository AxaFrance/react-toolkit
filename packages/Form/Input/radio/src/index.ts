export { default as RadioItem } from './RadioItem';
export { default as RadioInput } from './RadioInput';
export { default as Radio } from './Radio';
export { default as RadioModes } from './RadioModes';
