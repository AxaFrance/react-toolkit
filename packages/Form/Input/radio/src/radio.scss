@import '@axa-fr/react-toolkit-core/src/common/scss/core.scss';

.af-form {
  &__radio {
    display: block;
    align-items: center;
    position: relative;
    border-bottom: 1px solid $color-silver;

    &:nth-last-child(1) {
      border-bottom: 0;
    }
  }

  &__radio-inline {
    display: inline-flex;
    align-items: center;
    margin-right: 1rem;
    position: relative;

    .af-form__label {
      padding: 0;
    }
  }

  &__radio,
  &__radio-inline {
    &--error {
      .af-form__label {
        color: $color-red-axa;

        &::before {
          border-color: $color-red-axa;
        }
      }
    }
    &--warning {
      .af-form__label {
        color: $color-orange-dark;

        &::before {
          border-color: $color-orange-dark;
        }
      }
    }

    &--disabled {
      .af-form__label {
        opacity: 0.4;
        cursor: not-allowed;
      }
    }

    .af-form__label {
      &::before {
        content: '';
        position: absolute;
        display: block;
        margin-right: 0.5em;
        width: 1.25rem;
        height: 1.25rem;
        background: $white;
        border: 1px solid $color-silver;
        border-radius: 100%;
        left: 0;
        top: 50%;
        transform: translate(0, -50%);
      }

      &::after {
        content: '';
        position: absolute;
        display: none;
        margin-right: 0.5em;
        width: 0.9rem;
        height: 0.9rem;
        background: $color-azur;
        border-radius: 100%;
        left: 3px;
        top: 50%;
        transform: translate(0, -50%);
      }
    }

    .af-form__input-radio {
      position: absolute;
      opacity: 0;

      &:checked + .af-form__label {
        &::after {
          display: block;
        }
      }

      &:focus + .af-form__label {
        color: $color-axa;

        &::before {
          border-color: $color-axa;
        }
      }
    }
  }

  &__radio-custom {
    display: inline-flex;
    align-items: center;
    margin: 0 0.3rem 0.3rem 0;
    text-align: center;
    position: relative;

    .af-form__label {
      padding: 0.6rem 2rem;
      background: $color-mercury;
      border: $color-gray;
      font-weight: normal;
      cursor: pointer;
      box-shadow: inset 0 -2px 0 0 $color-scorpion;
      margin: 0;
    }

    .af-form__description {
      padding: 0;
    }

    .af-form__input-radio {
      position: absolute;
      opacity: 0;

      &:checked + .af-form__label {
        background-color: $color-azur;
        color: $white;
        box-shadow: inset 0 -2px 0 0 $color-axa;
      }

      &:focus + .af-form__label {
        background-color: $color-azur;
        color: $white;
        box-shadow: 0 0 5px $color-axa;
      }
    }

    &--disabled {
      .af-form__label {
        cursor: not-allowed;
        opacity: 0.6;
        box-shadow: inset 0 -2px 0 0 $color-silver;
      }

      .af-form__description {
        opacity: 0.4;
      }
    }

    &--error {
      .af-form__label {
        background: $error-custom-bg;
        border: 1px solid $error-custom-border;
        color: $color-red-axa;
        box-shadow: inset 0 -2px 0 0 $color-red-axa;
      }
    }

    &--warning {
      .af-form__label {
        background: $color-alertDanger-bgColor;
        border: 1px solid $color-alertDanger-colorBorder;
        color: $color-orange-dark;
        box-shadow: inset 0 -2px 0 0 $color-orange-dark;
      }
    }

    &:hover {
      .af-form__label {
        background-color: $color-azur;
        border-color: $color-axa;
        color: $white;
        box-shadow: inset 0 -2px 0 0 $color-axa;
      }
    }
  }
}
