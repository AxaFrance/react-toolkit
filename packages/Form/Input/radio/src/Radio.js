import React from 'react';
import PropTypes from 'prop-types';
import {
  InputList,
  InputConstants as Constants,
  withInput,
  omit,
} from '@axa-fr/react-toolkit-form-core';
import RadioItem from './RadioItem';
import RadioModes from './RadioModes';

const omitProperties = omit(['mode', 'helpMessage', 'id']);

const Radio = ({
  isVisible,
  className,
  disabled,
  options,
  value,
  name,
  onBlur,
  onFocus,
  readOnly,
  classModifier,
  children,
  onChange,
  ...otherProps
}) =>
  options.map(option => {
    const isChecked = option.value === value;
    return (
      <RadioItem
        key={option.value}
        id={option.id}
        value={option.value}
        label={option.label}
        isChecked={isChecked}
        name={name}
        onChange={onChange}
        onBlur={onBlur}
        onFocus={onFocus}
        readOnly={readOnly}
        isVisible={isVisible}
        disabled={InputList.isDisabled(option, disabled)}
        className={className}
        classModifier={option.classModifier}
        {...omitProperties(otherProps)}>
        {children}
      </RadioItem>
    );
  });

const propTypes = {
  ...Constants.propTypes,
  options: PropTypes.array.isRequired,
  value: PropTypes.string,
  mode: PropTypes.oneOf([
    RadioModes.default,
    RadioModes.classic,
    RadioModes.inline,
  ]),
};
const defaultClassName = '';
const defaultProps = {
  ...Constants.defaultProps,
  value: '',
  classModifier: defaultClassName,
  mode: RadioModes.default,
};

const handlers = {
  onChange: ({ onChange, name, id }) => e => {
    onChange({
      value: e.value,
      name,
      id,
    });
  },
};

const props = ({ mode }) => {
  let className = '';
  switch (mode) {
    case RadioModes.classic:
      className = 'af-form__radio';
      break;
    case RadioModes.inline:
      className = 'af-form__radio-inline';
      break;
    default:
      className = 'af-form__radio-custom';
      break;
  }
  return { className };
};

const EnhancedComponent = withInput(
  defaultClassName,
  propTypes,
  defaultProps,
  handlers,
  props
)(Radio);

EnhancedComponent.propTypes = propTypes;
EnhancedComponent.defaultProps = defaultProps;

EnhancedComponent.Clone = InputList.Clone;
EnhancedComponent.displayName = Radio.name;

export default EnhancedComponent;
