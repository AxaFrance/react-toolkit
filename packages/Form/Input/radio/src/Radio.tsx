import React, { ComponentProps } from 'react';
import RadioItem from './RadioItem';
import RadioModes from './RadioModes';

type Option = import('@axa-fr/react-toolkit-core/dist/InputManager').Option;
interface RadioProps
  extends Omit<ComponentProps<typeof RadioItem>, 'id' | 'label'> {
  className?: string;
  classModifier?: string;
  options: Option[];
  mode?: string;
}

const Radio: React.FC<RadioProps> = ({
  className,
  classModifier,
  options,
  value = '',
  name,
  mode = RadioModes.default,
  children,
  disabled,
  ...otherProps
}) => {
  const classNameWithMode = getClassName(mode);
  const optionItems = options.map((option: Option) => {
    const isChecked = option.value === value;
    return (
      <RadioItem
        key={option.value}
        id={option.id}
        value={option.value}
        label={option.label}
        checked={isChecked}
        disabled={option.disabled || disabled}
        className={classNameWithMode}
        classModifier={classModifier}
        optionClassName={className}
        {...otherProps}>
        {children}
      </RadioItem>
    );
  });
  return <>{optionItems}</>;
};

const getClassName = (mode: string): string => {
  switch (mode) {
    case RadioModes.classic:
      return 'af-form__radio';
    case RadioModes.inline:
      return 'af-form__radio-inline';
    default:
      return 'af-form__radio-custom';
  }
};

export default Radio;
