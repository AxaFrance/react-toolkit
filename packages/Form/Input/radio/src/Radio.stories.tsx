import { Meta, Story } from '@storybook/react';
import React, { ComponentProps } from 'react';
import { Radio, RadioModes } from '.';
import readme from '../README.md';
import { Option } from '../../../../core/src/InputManager';

export default {
  title: 'Form/Input/RadioInput/Radio',
  component: Radio,
  parameters: {
    readme: {
      sidebar: readme,
    },
  },
} as Meta;

export const RadioStory: Story<ComponentProps<typeof Radio>> = (args) => (
  <form className="af-form" name="myform">
    <div className="row">
      <div className="col-md-10">
        <Radio {...args} />
      </div>
    </div>
  </form>
);
const options = [
  { label: 'For fun', value: '1' },
  { label: 'For work', value: '2', id: 'customId' },
  { label: 'For drink', value: '3' },
  { label: 'For the life', value: '4', disabled: true },
] as Option[];

const modes = [RadioModes.classic, RadioModes.default, RadioModes.inline];
RadioStory.storyName = 'Radio';
RadioStory.args = {
  name: 'placeName',
  id: '',
  options,
  mode: RadioModes.default,
  value: '',
  placeholder: 'Paris',
  readOnly: false,
  disabled: false,
  classModifier: '',
  className: '',
};
RadioStory.argTypes = {
  mode: { control: { type: 'select', options: modes } },
  onChange: { action: 'onChange' },
};
