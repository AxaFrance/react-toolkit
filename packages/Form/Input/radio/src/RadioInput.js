import React from 'react';
import PropTypes from 'prop-types';

import {
  Field,
  FieldConstants as Constants,
} from '@axa-fr/react-toolkit-form-core';
import { InputManager } from '@axa-fr/react-toolkit-core';

import Radio from './Radio';
import RadioModes from './RadioModes';

const propTypes = {
  ...Constants.propTypes,
  options: PropTypes.array.isRequired,
  value: PropTypes.string,
  mode: PropTypes.oneOf([
    RadioModes.default,
    RadioModes.classic,
    RadioModes.inline,
  ]),
};

const defaultProps = {
  ...Constants.defaultProps,
  value: '',
  mode: RadioModes.default,
};

const RadioInput = props => {
  const {
    mode,
    messageType,
    message,
    className,
    classModifier,
    id,
    name,
    isVisible,
    options,
    classNameContainerLabel,
    classNameContainerInput,
    label,
    forceDisplayMessage,
    value,
    onChange,
    readOnly,
    disabled,
    children,
    ...otherProps
  } = props;
  if (!isVisible) {
    return null;
  }
  let rowModifier = classModifier;
  if (mode === RadioModes.classic) {
    rowModifier += ' label-top';
  }
  const newOptions = InputManager.getOptionsWithId(options);
  const firstId = InputManager.getFirstId(newOptions);
  return (
    <Field
      label={label}
      name={name}
      id={firstId}
      message={message}
      messageType={messageType}
      isVisible={isVisible}
      forceDisplayMessage={forceDisplayMessage}
      className={className}
      classModifier={rowModifier}
      classNameContainerLabel={classNameContainerLabel}
      classNameContainerInput={classNameContainerInput}>
      <Radio
        name={name}
        id={id}
        options={newOptions}
        value={value}
        mode={mode}
        onChange={onChange}
        readOnly={readOnly}
        disabled={disabled}
        classModifier={classModifier}
        {...otherProps}
      />
      {children}
    </Field>
  );
};

RadioInput.propTypes = propTypes;
RadioInput.defaultProps = defaultProps;

export default RadioInput;
