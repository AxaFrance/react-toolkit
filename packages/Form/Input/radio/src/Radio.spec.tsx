import { shallow } from 'enzyme';
import React, { ComponentProps } from 'react';
import Radio from './Radio';
import RadioItem from './RadioItem';
import { Option } from '../../../../core/src/InputManager';

describe('Radio', () => {
  describe('template', () => {
    let args: ComponentProps<typeof Radio>;

    beforeEach(() => {
      const options = [
        { label: 'For fun', value: '1' },
        { label: 'For work', value: '2', id: 'customId' },
        { label: 'For drink', value: '3' },
        { label: 'For the life', value: '4', disabled: true },
      ] as Option[];
      args = { options };
    });

    const Act = () => shallow(<Radio {...args} />);

    it('should have correct html rendered', () => {
      const radioComponent = Act();
      expect(radioComponent.html).toMatchSnapshot();
    });

    describe('isSelected', () => {
      it.each`
        value        | expected
        ${'1'}       | ${true}
        ${'2'}       | ${false}
        ${null}      | ${false}
        ${undefined} | ${false}
      `(
        'should set checked to $expected when selectedValue is $value',
        ({ value, expected }) => {
          args.value = value;
          const radioComponent = Act();
          expect(
            radioComponent
              .find(RadioItem)
              .findWhere((node) => node.prop('value') === '1')
              .prop('checked')
          ).toBe(expected);
        }
      );
    });
  });
});
