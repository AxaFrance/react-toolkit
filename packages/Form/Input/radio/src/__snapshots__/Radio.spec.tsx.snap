// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio template should have correct html rendered 1`] = `[Function]`;
